src/normalize.js

```diff
@@ -1,42 +1,60 @@
+const cmp = (a, b) => {
+  if (a < b) {
+    return -1;
+  } else if (a > b) {
+    return 1;
+  } else {
+    return 0;
+  }
+};
+
 /**
  * Compare scopes a and b to see which comes first if sorted
- * Such that 'a*' comes before 'a', but otherwise normal order.
+ * Such that 'a*' comes before 'a', but otherwise normal order.  In other words,
+ * the trailing * is considered a special character that sorts *before* the end of
+ * string.
+ *
+ * In particular, for a given prefix P that does not end in *, it will produce
+ * the following contiguous segments of the set of scopes beginning with P (if
+ * such scopes are present):
+ *
+ *  - P*
+ *  - P
+ *  - P<anything>
+ *
+ * In the case where P itself ends in *, the * in "P<anything>" is not a trailing star
+ * and is thus not treated specially, so ordering is lexical.
  *
  * Example: ['*', '', 'a*', 'a', 'a(', 'aa', 'b'] is a list sorted as such.
  *
  * This sort order is useful for normalizing scopes, since it puts the "interesting"
  * scopes (those with a final `*`) up front.
  */
 export const scopeCompare = (a, b) => {
```

Comment:
In an effort to understand this section, I wrote it a different way


```js
export const scopeCompare = (a, b) => {
  let astar = a.endsWith('*');
  let bstar = b.endsWith('*');

  a = astar ? a.slice(0, -1) : a;
  b = bstar ? b.slice(0, -1) : b;

  if (astar !== bstar && a === b) {
    return astar ? -1 : 1;
  }
  return cmp(a, b);
};
```

I believe they are equivalent. Does this one seem easier to understand or harder to understand to you? I'm honestly not sure which I prefer.

Reply:
Yes, the two are equivalent, and I like your version better.